### ISSUE TITLE
Advanced usage of set for ignoring state changes

### ISSUE THREAD

**Issue by funwithtriangles** [charkour/zundo#9]
It would be nice if we could ignore state on a granular level, for any state change. 

Perhaps you have a piece of state that changes very often and you only want to store those changes occasionally. One example would be the a rotation property on a 3D object. Perhaps you have have an app where you must rotate 3D objects. These update every frame but you only want to store the beginning and end state of the user's rotation. It might be convenient to have this information in the store, but without ignoring most of the changes, you'll have to click "undo" 100 times just to get back to the beginning of the rotation!

This could be solved by adding an extra argument to the step function:

```javascript
const useStoreWithUndo = create<StoreState>(set => ({
  rotationX: 0,
  ...
  setRotation: (rotationX) => set(state => ({ rotationX })),
  setRotationWithIgnore: (rotationX) => set(state => ({ rotationX }), true), // Set second arg to true to ignore this change
}));
```

In a more complex scenario, you may want to use current state to determine whether to set this argument:

```javascript
const useStoreWithUndo = create<StoreState>((set, get) => ({
  ...
  setRotation: (rotationX) => {
    const shouldIgnore = get().isDragging
    return set(state => ({ rotationX }), shouldIgnore)
  } 
}));
```

I'm not entirely sure if this is the best API but thought I'd share my thoughts on the problem and a possible solution.

**Comment by charkour** [comment]
Thanks for this issue request. I think the problem you describe and the solution are fantastic. I agree that coming up with a good API for this would be difficult but your initial suggestions are a great starting point.

So in the first code snippet, `setRotationWithIgnore` would be called while the object is rotating, and at the end of the movement, `setRotation` would be called? Perhaps instead the API could be `setRotation: (rotationX, shouldIgnore) => set(state => ({ rotationX }), shouldIgnore),` that way you don't have to repeat similar logic in your store. What do you think about that?

As for the second code snippet, programmatically setting `shouldIgnore` I think it a good pattern to use, similar to the suggestion I had above so logic isn't repeated.

So thinking through this, I like the API you suggest and we can go ahead and implement it. Would you like to make a PR for this feature?

**Comment by funwithtriangles** [comment]
Yes I was just putting a super simple example in first to get the point across. Definitely think you could either add an extra argument to your setter function (as you have shown) or use `get` to do the logic inside of the function without the need for that extra argument.

I'd love to have a go at doing a PR for this :) The only other possible thing I was thinking for the API was to pass in a settings object rather than a single argument. Something like this: 

`set(state => ({ rotationX }), { shouldIgnore: true })`

It depends if we think there will be need for many more options in the future or not. I'm not entirely sure on my preference.

**Comment by charkour** [comment]
I do like the idea of a settings object. Let's go with that!

**Comment by funwithtriangles** [comment]
Just so you know, I do have a need for this and would still like to do the PR, but it's something I'll pick up in a few weeks time when it's relevant for the project I'm on.

**Comment by charkour** [comment]
Sounds good, thanks for letting me know!

I'll review the PR once it is ready.

**Comment by charkour** [comment]
Oh yes, overriding the set foundation could cause issues. I assumed that `zundo` would be used at the base but I'm thinking that was a wrong assumption. 

Maybe I can release v2 that doesn't override the `set` function or just provide better typings.

**Comment by charkour** [comment]
I can review your code later this weekend. Thanks!
